File: server/spec/factories/ai/mcp_apps.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :ai_mcp_app, class: "Ai::McpApp" do
    account
    sequence(:name) { |n| "MCP App #{n}" }
    description { "A test MCP application" }
    app_type { "custom" }
    status { "draft" }
    html_content { "<div>Hello World</div>" }
    csp_policy { {} }
    sandbox_config { {} }
    input_schema { {} }
    output_schema { {} }
    metadata { {} }
    version { "1.0.0" }

    trait :draft do
      status { "draft" }
    end

    trait :published do
      status { "published" }
    end

    trait :archived do
      status { "archived" }
    end

    trait :custom do
      app_type { "custom" }
    end

    trait :template do
      app_type { "template" }
    end

    trait :system do
      app_type { "system" }
    end

    trait :with_schema do
      input_schema do
        {
          "type" => "object",
          "required" => ["name"],
          "properties" => {
            "name" => { "type" => "string" },
            "count" => { "type" => "integer" }
          }
        }
      end
      output_schema do
        {
          "type" => "object",
          "properties" => {
            "result" => { "type" => "string" }
          }
        }
      end
    end

    trait :with_csp do
      csp_policy do
        {
          "script-src" => "'self'",
          "style-src" => "'self' 'unsafe-inline'"
        }
      end
    end
  end
end

File: frontend/src/features/ai/mcp-apps/types/mcpApps.ts
// MCP App Types
export type McpAppType = 'custom' | 'template' | 'system';
export type McpAppStatus = 'draft' | 'published' | 'archived';

// MCP App (list/summary shape)
export interface McpApp {
  id: string;
  account_id: string;
  name: string;
  description: string | null;
  app_type: McpAppType;
  status: McpAppStatus;
  version: string;
  created_by_id: string | null;
  input_schema: Record<string, unknown>;
  output_schema: Record<string, unknown>;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

// MCP App (detailed shape, includes html_content, csp, sandbox)
export interface McpAppDetailed extends McpApp {
  html_content: string | null;
  csp_policy: Record<string, unknown>;
  sandbox_config: Record<string, unknown>;
  instance_count: number;
}

// Render result
export interface McpAppRenderResult {
  html: string;
  instance_id: string;
  csp_headers: Record<string, string>;
  sandbox_attrs: string;
}

// Process input result
export interface McpAppProcessResult {
  response: Record<string, unknown>;
  state_update: Record<string, unknown> | null;
}

// API params
export interface McpAppFilterParams {
  status?: McpAppStatus;
  app_type?: McpAppType;
  search?: string;
}

export interface CreateMcpAppParams {
  name: string;
  description?: string;
  app_type: McpAppType;
  status?: McpAppStatus;
  html_content?: string;
  version?: string;
  csp_policy?: Record<string, unknown>;
  sandbox_config?: Record<string, unknown>;
  input_schema?: Record<string, unknown>;
  output_schema?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface UpdateMcpAppParams extends Partial<CreateMcpAppParams> {
  id: string;
}

export interface RenderMcpAppParams {
  id: string;
  session_id?: string;
  context?: Record<string, unknown>;
}

export interface ProcessMcpAppInputParams {
  id: string;
  instance_id: string;
  input_data: Record<string, unknown>;
}
